### ISSUE TITLE
[clang-format] regression from clang-format v13 

### ISSUE THREAD

**Issue by paulhoad** [llvm/llvm-project#53567]
The following source

```
namespace A {

template <int N> struct Foo<char[N]> {
  void foo() { std::cout << "Bar"; }
}; // namespace A
```

is formatted as:

```
namespace A {

template <int N> struct Foo<char[N]>{void foo(){std::cout << "Bar";
}
}
; // namespace A
```

using clang-format version 15.0.0 (https://github.com/llvm/llvm-project.git ee4ba9f3a1
82397c60e17f14823eb22b2e5831c7)

Using the following .clang-format file

`BasedOnStyle: LLVM`

**Comment by mydeveloperday** [comment]
possibly caused by the presence of 

`[N]`

given

```
namespace A {

template <int N> struct Foo<char[N]>{void foo(){std::cout << "Bar";
}
}
; // namespace A

namespace B {

template <int N> struct Foo<char> {
  void foo() { std::cout << "Bar"; }
}; // namespace A
```

**Comment by mkurdej** [comment]
With clang-format from 2022-01-05, commit 5109737c:
```
clang-format-main-20220105-T1352-5109737c.exe --style=file test.cpp
namespace A {

template <int N> struct Foo<char[N]> {
  void foo() { std::cout << "Bar"; }
};

} // namespace A
```

And with clang-format from 2022-01-10, commit 5c2e7c9ca043:
```
clang-format-main-20220110-5c2e7c9ca043.exe --style=file test.cpp
namespace A {

template <int N> struct Foo<char[N]>{void foo(){std::cout << "Bar";
}
}
;

} // namespace A
```

I think it's the commit https://github.com/llvm/llvm-project/commit/5c2e7c9ca043d92bed75b08e653fb47c384edd13 that provoked that.

**Comment by mkurdej** [comment]
Commit https://github.com/llvm/llvm-project/commit/72e29caf039fd81bc6948e16d6f71d1581615469 also fixed a regression on it.
I think that the solution would be to call `tryToParseLambda` only if the previous token is `<` (template opener) or `,` (template argument comma) *when we are in the template context*.

**Comment by mydeveloperday** [comment]
https://reviews.llvm.org/D118911

**Comment by mydeveloperday** [comment]
Fixed by commit 23fc20e06c088acff81a06ad546a848bee083051

**Comment by mydeveloperday** [comment]
@tstellar could this possibly be cherry-picked back into the 14 branch?

**Comment by mydeveloperday** [comment]
ReOpened until this is backported

**Comment by mkurdej** [comment]
> How did you get to the commit so quickly! (are you keeping a copy of the .exe at every commit?)

Not every one but a bunch of them (like one per week + all releases). That saves time!

**Comment by mydeveloperday** [comment]
I must start doing that, it's an excellent idea... there was me ploughing through git checkouts from Dec/Jan and you were on it already!!

**Comment by tstellar** [comment]
Merged: 592367ab7ea8c70eb26f2a6a7ddeb05b50322545
